docs/blueprints/EP035.rst

```diff
@@ -0,0 +1,208 @@
+:Type: Standards Track
+
+*******************************
+EP035 - Interruption Management
+*******************************
+
+########
+Abstract
+########
+
+This blueprint details a mechanism for NApps providing services
+to receive information on interruptions,
+and provide information on the affected services back
+to the source of the interruption.
+Additionally, this blueprint details how NApps which produce
+interruptions should signal them in order to take advantage
+of this new mechanism.
+
+##########
+Motivation
+##########
+
+In order for NApps which produce service interruptions
+to receive information on what services they are affecting,
+there needs to be a two way communication channel between
+the service provider and the interruption source.
+The kytos event bus provides only one way communication,
+making it unsuitable for this task.
+
+#############
+Specification
+#############
+
+The following is a set of proposed changes for
+how to handle interruptions in the future,
+in order to have visibility on there causes and effects.
+
+Kytos Core Changes
+##################
+
+In order for services to identify which interruptions are affecting the devices
+they use, a mechanism needs to be added to the device representation for
+identify interruption causes.
+
+
+Extend ``status_funcs`` to produce a ``status_reason``
+------------------------------------------------------
+
+Currently, when an object is queried about its ``status``,
+the reason for that status is unknown.
+What we propose here is adding a new property to
+``GenericEntity`` called ``status_reason``,
+and adding in ``status_reason_func``
+to produce values for ``status_reason``.
+The produced ``status_reason`` will be a set of strings,
+so all possible reasons a device might be disable
+can be easily represented.
+
+Example usage:
+
+.. code-block:: python3
+
+    from kytos.core.switch import Switch
+    maintenance_switches = set()
+
+    def switch_maintenance_status(switch):
+        if switch.id in maintenance_switches:
+            return EntityStatus.DOWN
+        return EntityStatus.UP
+
+    Switch.register_status_func('maintenance', switch_maintenance_status)
+
+    def switch_maintenance_reason(switch):
+        if switch.id in maintenance_switches:
+            return {'maintenance'}
+        return set()
+
+    Switch.register_status_reason_func('maintenance', switch_maintenance_reason)
+
+
+New Topology REST Endpoints
+###########################
+
+The following set of new endpoints are intended to allow
+for previewing the affects of adding in interruptions.
+
+
+REST Endpoint ``GET /topology/v3/preview_interruption``
+-------------------------------------------------------
+
+For all know service providers,
+previews affected if the specified interruption occurs now.
+
+Expected Input:
+
+.. code-block:: http
+
+    GET /topology/v3/preview_interruption?switch=<switch_id>&link=<link_id>&interface=<interface_id>&type=<interruption_type>
```

Comment:
It's typical for GET to take the filter params on its url query args. For this case though, can we instead provide the filters in the request body, and ideally following closely the same shape format that `maitenance` uses and `topology.interruptions.start`? Something like:

```
 {
        "switches": [<switch id>],
        "interfaces": [<interface id>],
        "links": [<link id>],
 }
```

That way, pretty much the payload can be reused to schedule a maintenance or an interruption once a network engineer is happy with the interruption preview result impact. Also, when filtering/querying for an interruption shouldn't we provide a filter for a service provider and which of its services instead of an interruption type?